tools/tune/multiband_drc/iir_gen_quant_coefs.m

```diff
@@ -0,0 +1,101 @@
+function [emp_coefs, deemp_coefs] = iir_gen_quant_coefs(params);
+
+stage_gain = params.stage_gain;
+stage_ratio = params.stage_ratio;
+anchor_freq = params.anchor;
+
+emp = cell(1, 2);
+deemp = cell(1, 2);
+% Generate the coefficients of (de)emphasis for the 1-st stage of biquads
+[emp(1), deemp(1)] = emp_deemp_stage_biquad(stage_gain, anchor_freq,
+					    anchor_freq / stage_ratio);
+
+anchor_freq = anchor_freq / (stage_ratio * stage_ratio);
+
+% Generate the coefficients of (de)emphasis for the 2-nd stage of biquads
+[emp(2), deemp(2)] = emp_deemp_stage_biquad(stage_gain, anchor_freq,
+                                            anchor_freq / stage_ratio);
+
+% Print emp and deemp
+emp
+deemp
+
+% Convert the coefficients to values usable with SOF
+emp_coefs = iir_coef_quant(emp);
+deemp_coefs = iir_coef_quant(deemp);
+
+end
+
+function [emp, deemp] = emp_deemp_stage_biquad(gain, f1, f2);
+
+[z1, p1] = emp_stage_roots(gain, f1);
+[z2, p2] = emp_stage_roots(gain, f2);
+
+b0 = 1;
+b1 = -(z1 + z2);
+b2 = z1 * z2;
+a0 = 1;
+a1 = -(p1 + p2);
+a2 = p1 * p2;
+
+% Gain compensation to make 0dB at 0Hz
+alpha = (a0 + a1 + a2) / (b0 + b1 + b2);
```

Comment:
I think this causes the saturation that you reported. I'm scaling the biquads so that max. gain at any frequency is -6 dB. The gain is restored in the last shift & gain. Please be careful with your new IIR core, an overflow or even saturation could make IIR state variables instable.

Reply:
IIR scaling is AFAIK quite empirical, there's different norms but some could be too optimistical and some too pessimistical (lost SNR with low word lengths like 32 bit). I've used fast chirp signal sometimes, they seem to cause large internal levels. Also random noise that is clipped with gain and saturation. That said, I'm not sure I've checked those for EQ but need to do with with the reported coefficients and the scaling procedure for IIR EQ filters.

Reply:
Thanks Seppo. You have provided many information for me to utilize iir_df2t() more wisely.
After the digestion and some experiments, I changed my emphasis and deemphasis filter coefficients with the following rules:
1. apply https://github.com/thesofproject/sof/pull/3684 to have sat_int32 "tmp" variable.
2. for emphasis filter, each stage gain (alpha) is larger than 1. The gain should be pushed and aggregated in the last stage (and decompose to shift and gain).
3. for deemphasis filter, each stage gain (beta) is smaller than 1 (which is not encouraged). Then beta will be multiplied directly to b0, b1, and b2, and keep each stage gain is 1.0, in order to achieve a healthier biquad scaling.

With such adjustments, we don't need to customize iir_df2t function anymore. Those 2 filters look stable with my test streams.

Reply:
Great, good that the filters can be scaled. I still wonder what happens with emphasis filter if it's fed with high frequency sine wave, I recall it's the first operation in MDRC. The IIRs will internally perform OK with new scaling but the output will be saturated. The high frequency emphasis could be done with low frequencies attenuate instead. Since the internal word length of DRC is 32 bits there's plenty of dynamics left to do that. The de-emphasis filter can have the needed high gain to restore the level.

Reply:
@johnylin76 is this complete?

Reply:
Yes, thanks.